lib/puppet/module.rb

```diff
@@ -456,7 +456,7 @@ def subpath(type)
 
   def assert_validity
     if !Puppet::Module.is_module_directory_name?(@name) && !Puppet::Module.is_module_namespaced_name?(@name)
-      raise InvalidName, "Invalid module name #{@name}; module names must be alphanumeric (plus '-'), not '#{@name}'"
+      raise InvalidName, "Invalid module name #{@name}; Module names must begin with a lowercase letter, and can include; Uppercase letters , Lowercase letters, Digits and Underscores (regex '\A[a-z][a-z0-9_]*\Z'), not '#{@name}'"
```

Comment:
Suggest quoting this part: `'#{@name}`  - since it can be some word that renders the sentence incomprehensible - say the name is "unless bears poop in the forest" ;-)

Then, the last bit `not '#{@name}'" can be dropped as it is the same thing again - don't think repeating it adds any value.